tidy3d/components/data/sim_data.py

```diff
@@ -978,9 +982,158 @@ def make_adjoint_sim(
             grid_spec_adj = grid_spec_original.updated_copy(wavelength=wavelength_original)
             sim_adj_update_dict["grid_spec"] = grid_spec_adj
 
-        return sim_original.updated_copy(**sim_adj_update_dict)
+        return sim_original.updated_copy(**sim_adj_update_dict), post_norm_amps
+
+    def process_adjoint_sources(
+        self, sources_adj_dict: dict[str, list[Source]]
+    ) -> tuple[list[Source], xr.DataArray]:
+        """Process mapping of monitor name to adjoint sources to adj srcs and post-norm amps."""
+
+        # number of monitors contributing to adjoint sources
+        num_adj_src_monitors = len(sources_adj_dict)
+
+        # no adjoint sources, gradient is 0
+        if num_adj_src_monitors == 0:
+            log.warning(
+                "Something unexpected happened. There are No adjoint sources, "
+                " yet the adjoint pipeline was triggered. No gradient will be computed "
+                "with respect to simulation data output. If you receive this, please file a bug "
+                "report on the tidy3d github repository with as much information as "
+                "you can provide about your setup."
+            )
+            return [], None
```

Comment:
When would this happen? Can this happen when then gradient w.r.t. a specific monitor is exactly zero?

Reply:
hmmm. yea I dont know. so if we do something like `y = 0*x`, I'm not sure if autograd still traces `y`? in any case, it's unexpected, but I dont think we need to error? but the user might want to know what's going on so I felt a warning would be sufficient to catch this weird edge case.

Reply:
Yeah no I agree with not erroring, I was just confused because this tells the user to submit a bug report 😄 Maybe the text should be changed to something like „If this is unexpected …“?

Reply:
good point, I'll update it